hackernews.py

```diff
@@ -1,85 +1,61 @@
-        self.firebase = firebase.FirebaseApplication('https://hacker-news.firebaseio.com/', None)
+from aiohttp import ClientSession
+from firebase import firebase
+from itertools import chain
 
-    def top_stories(self):
+
+class HnApi:
+    def __init__(self, loop=None, session=None):
+        self.loop = loop if loop else asyncio.get_event_loop()
+        self.session = session if session else ClientSession(loop=self.loop)
+        self.api_root = 'https://hacker-news.firebaseio.com/'
+        self.firebase = firebase.FirebaseApplication(
+            'https://hacker-news.firebaseio.com/', None)
+
+    def get_top_stories(self):
         """Retrieves the top stories"""
         try:
-            return self.firebase.get('/v0/topstories',None)
-        except:
+            return self.firebase.get('/v0/topstories', None)
+        except requests.exceptions.RequestException:
             print("Couldn't fetch top stories")
-    def get_item(self,item):
+            raise
+
+    def get_item(self, item):
         try:
-            return self.firebase.get('/v0/item/'+str(item),None)
-        except Exception as E:
-            print("item "+str(item)+" not found")
-            print(E)
-    async def _get_item(self,item):
-        async with self._session.get(self._api_root+'/v0/item/'+str(item)+'.json') as response:
+            return self.firebase.get(f'/v0/item/{item}', None)
+        except requests.exceptions.RequestException as e:
+            print("item '{item}' not found")
+            raise
+
+    async def async_get_item(self, item):
+        item_url = f"{self.api_root}v0/item/{item}.json"
+        async with self.session.get(item_url) as response:
             f = await response.text()
-            obj = json.loads(f)
-            return obj
-            
+            return json.loads(f)
+
     def get_comments(self,item):
         """Nasty peice of works - BFS without history"""
-        queue = [item]
-        res =[]
-        while queue:
-            execution = queue.copy()
-            tasks = list(map(lambda x: self._get_item(x) if type(x) == type(1) else x, execution))
-            queue = []
-            finished, unfinished = self.loop.run_until_complete(
+        new_items = [item]
+        pending = []
+        results = []
+        while new_items or pending:
+            tasks = chain(map(self.async_get_item, new_items), pending)
+            done, pending = self.loop.run_until_complete(
                 asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED))
-            for task in finished:
+
+            new_items = []
+            for task in done:
                 r = task.result()
                 if r["type"] == "comment" and "text" in r:
-                    res.append((r["id"],r["text"]))
-                try:
-                    queue += r["kids"]
-                except KeyError:
-                    pass
-            queue += unfinished
-        return res
-    def __del__(self):
```

Comment:
I am using __del__ here to prevent memory leaks without us having to go and manually clean up the mess